src/fs/base/tree.ts

```diff
@@ -0,0 +1,137 @@
+import pathUtil from '../path'
+import { Links, SimpleTree, SimpleFile, SemVer, NonEmptyPath } from '../types'
+import check from '../types/check'
+import { CID, FileContent } from '../../ipfs'
+
+abstract class BaseTree implements SimpleTree {
+
+  version: SemVer
+
+  constructor(version: SemVer) {
+    this.version = version
+  }
+
+  async ls(path: string): Promise<Links> {
+    const dir = await this.get(path)
+    if (dir === null) {
+      throw new Error("Path does not exist")
+    } else if (check.isSimpleFile(dir)) {
+      throw new Error('Can not `ls` a file')
+    }
+    return dir.getLinks()
+  }
+
+  async mkdir(path: string): Promise<this> {
+    const exists = await this.pathExists(path)
+    if (exists) {
+      throw new Error(`Path already exists: ${path}`)
```

Comment:
Had this convo with @expede when I was first making the api, and we decided to throw an error (similar to how `mkdir` works in unix), since it gives the user more feedback as to what's happening under the hood. I was thinking it might be good to add a flag that lets you toggle that behavior